File: sdks/csharp/src/NovaBilling/Webhooks/WebhooksClient.cs
using NovaBilling.Core;

namespace NovaBilling;

public partial class WebhooksClient : IWebhooksClient
{
    private RawClient _client;

    internal WebhooksClient(RawClient client)
    {
        _client = client;
    }

    /// <summary>
    /// Receives payment event notifications from Paystack. The signature is verified using HMAC-SHA512 with the provider's secret key. On success, updates the payment/invoice status and sends customer notifications.
    /// </summary>
    /// <example><code>
    /// await client.Webhooks.WebhooksControllerPaystackAsync(
    ///     new WebhooksControllerPaystackRequest { PaystackSignature = "x-paystack-signature" }
    /// );
    /// </code></example>
    public async Task WebhooksControllerPaystackAsync(
        WebhooksControllerPaystackRequest request,
        RequestOptions? options = null,
        CancellationToken cancellationToken = default
    )
    {
        var _headers = await new NovaBilling.Core.HeadersBuilder.Builder()
            .Add("x-paystack-signature", request.PaystackSignature)
            .Add(_client.Options.Headers)
            .Add(_client.Options.AdditionalHeaders)
            .Add(options?.AdditionalHeaders)
            .BuildAsync()
            .ConfigureAwait(false);
        var response = await _client
            .SendRequestAsync(
                new JsonRequest
                {
                    BaseUrl = _client.Options.BaseUrl,
                    Method = HttpMethod.Post,
                    Path = "webhooks/paystack",
                    Headers = _headers,
                    Options = options,
                },
                cancellationToken
            )
            .ConfigureAwait(false);
        if (response.StatusCode is >= 200 and < 400)
        {
            return;
        }
        {
            var responseBody = await response.Raw.Content.ReadAsStringAsync();
            throw new NovabillingApiApiException(
                $"Error with status code {response.StatusCode}",
                response.StatusCode,
                responseBody
            );
        }
    }

    /// <summary>
    /// Receives payment event notifications from Flutterwave. Verified using the verif-hash header against the configured encryption key.
    /// </summary>
    /// <example><code>
    /// await client.Webhooks.WebhooksControllerFlutterwaveAsync(
    ///     new WebhooksControllerFlutterwaveRequest()
    /// );
    /// </code></example>
    public async Task WebhooksControllerFlutterwaveAsync(
        WebhooksControllerFlutterwaveRequest request,
        RequestOptions? options = null,
        CancellationToken cancellationToken = default
    )
    {
        var _headers = await new NovaBilling.Core.HeadersBuilder.Builder()
            .Add("verif-hash", request.VerifHash)
            .Add(_client.Options.Headers)
            .Add(_client.Options.AdditionalHeaders)
            .Add(options?.AdditionalHeaders)
            .BuildAsync()
            .ConfigureAwait(false);
        var response = await _client
            .SendRequestAsync(
                new JsonRequest
                {
                    BaseUrl = _client.Options.BaseUrl,
                    Method = HttpMethod.Post,
                    Path = "webhooks/flutterwave",
                    Headers = _headers,
                    Options = options,
                },
                cancellationToken
            )
            .ConfigureAwait(false);
        if (response.StatusCode is >= 200 and < 400)
        {
            return;
        }
        {
            var responseBody = await response.Raw.Content.ReadAsStringAsync();
            throw new NovabillingApiApiException(
                $"Error with status code {response.StatusCode}",
                response.StatusCode,
                responseBody
            );
        }
    }

    /// <summary>
    /// Receives payment callback notifications from DPO Group (DirectPay Online). Verifies the transaction token status and updates payment accordingly.
    /// </summary>
    /// <example><code>
    /// await client.Webhooks.WebhooksControllerDpoAsync();
    /// </code></example>
    public async Task WebhooksControllerDpoAsync(
        RequestOptions? options = null,
        CancellationToken cancellationToken = default
    )
    {
        var _headers = await new NovaBilling.Core.HeadersBuilder.Builder()
            .Add(_client.Options.Headers)
            .Add(_client.Options.AdditionalHeaders)
            .Add(options?.AdditionalHeaders)
            .BuildAsync()
            .ConfigureAwait(false);
        var response = await _client
            .SendRequestAsync(
                new JsonRequest
                {
                    BaseUrl = _client.Options.BaseUrl,
                    Method = HttpMethod.Post,
                    Path = "webhooks/dpo",
                    Headers = _headers,
                    Options = options,
                },
                cancellationToken
            )
            .ConfigureAwait(false);
        if (response.StatusCode is >= 200 and < 400)
        {
            return;
        }
        {
            var responseBody = await response.Raw.Content.ReadAsStringAsync();
            throw new NovabillingApiApiException(
                $"Error with status code {response.StatusCode}",
                response.StatusCode,
                responseBody
            );
        }
    }

    /// <summary>
    /// Receives Instant Payment Notifications (IPN) from PayU South Africa. Updates payment status based on the transaction state.
    /// </summary>
    /// <example><code>
    /// await client.Webhooks.WebhooksControllerPayuAsync();
    /// </code></example>
    public async Task WebhooksControllerPayuAsync(
        RequestOptions? options = null,
        CancellationToken cancellationToken = default
    )
    {
        var _headers = await new NovaBilling.Core.HeadersBuilder.Builder()
            .Add(_client.Options.Headers)
            .Add(_client.Options.AdditionalHeaders)
            .Add(options?.AdditionalHeaders)
            .BuildAsync()
            .ConfigureAwait(false);
        var response = await _client
            .SendRequestAsync(
                new JsonRequest
                {
                    BaseUrl = _client.Options.BaseUrl,
                    Method = HttpMethod.Post,
                    Path = "webhooks/payu",
                    Headers = _headers,
                    Options = options,
                },
                cancellationToken
            )
            .ConfigureAwait(false);
        if (response.StatusCode is >= 200 and < 400)
        {
            return;
        }
        {
            var responseBody = await response.Raw.Content.ReadAsStringAsync();
            throw new NovabillingApiApiException(
                $"Error with status code {response.StatusCode}",
                response.StatusCode,
                responseBody
            );
        }
    }

    /// <summary>
    /// Receives IPN (Instant Payment Notification) callbacks from Pesapal. Fetches transaction status using the OrderTrackingId and updates payment.
    /// </summary>
    /// <example><code>
    /// await client.Webhooks.WebhooksControllerPesapalAsync();
    /// </code></example>
    public async Task WebhooksControllerPesapalAsync(
        RequestOptions? options = null,
        CancellationToken cancellationToken = default
    )
    {
        var _headers = await new NovaBilling.Core.HeadersBuilder.Builder()
            .Add(_client.Options.Headers)
            .Add(_client.Options.AdditionalHeaders)
            .Add(options?.AdditionalHeaders)
            .BuildAsync()
            .ConfigureAwait(false);
        var response = await _client
            .SendRequestAsync(
                new JsonRequest
                {
                    BaseUrl = _client.Options.BaseUrl,
                    Method = HttpMethod.Post,
                    Path = "webhooks/pesapal",
                    Headers = _headers,
                    Options = options,
                },
                cancellationToken
            )
            .ConfigureAwait(false);
        if (response.StatusCode is >= 200 and < 400)
        {
            return;
        }
        {
            var responseBody = await response.Raw.Content.ReadAsStringAsync();
            throw new NovabillingApiApiException(
                $"Error with status code {response.StatusCode}",
                response.StatusCode,
                responseBody
            );
        }
    }

    /// <summary>
    /// Receives event notifications from Stripe (e.g. checkout.session.completed, payment_intent.succeeded). Verified using the stripe-signature header with the configured webhook secret.
    /// </summary>
    /// <example><code>
    /// await client.Webhooks.WebhooksControllerStripeAsync(
    ///     new WebhooksControllerStripeRequest { StripeSignature = "stripe-signature" }
    /// );
    /// </code></example>
    public async Task WebhooksControllerStripeAsync(
        WebhooksControllerStripeRequest request,
        RequestOptions? options = null,
        CancellationToken cancellationToken = default
    )
    {
        var _headers = await new NovaBilling.Core.HeadersBuilder.Builder()
            .Add("stripe-signature", request.StripeSignature)
            .Add(_client.Options.Headers)
            .Add(_client.Options.AdditionalHeaders)
            .Add(options?.AdditionalHeaders)
            .BuildAsync()
            .ConfigureAwait(false);
        var response = await _client
            .SendRequestAsync(
                new JsonRequest
                {
                    BaseUrl = _client.Options.BaseUrl,
                    Method = HttpMethod.Post,
                    Path = "webhooks/stripe",
                    Headers = _headers,
                    Options = options,
                },
                cancellationToken
            )
            .ConfigureAwait(false);
        if (response.StatusCode is >= 200 and < 400)
        {
            return;
        }
        {
            var responseBody = await response.Raw.Content.ReadAsStringAsync();
            throw new NovabillingApiApiException(
                $"Error with status code {response.StatusCode}",
                response.StatusCode,
                responseBody
            );
        }
    }
}

File: sdks/swift/Sources/Resources/Webhooks/WebhooksClient.swift
import Foundation

public final class WebhooksClient: Sendable {
    private let httpClient: HTTPClient

    init(config: ClientConfig) {
        self.httpClient = HTTPClient(config: config)
    }

    /// Receives payment event notifications from Paystack. The signature is verified using HMAC-SHA512 with the provider's secret key. On success, updates the payment/invoice status and sends customer notifications.
    ///
    /// - Parameter paystackSignature: Paystack HMAC-SHA512 signature
    /// - Parameter requestOptions: Additional options for configuring the request, such as custom headers or timeout settings.
    public func webhooksControllerPaystack(paystackSignature: String, requestOptions: RequestOptions? = nil) async throws -> Void {
        return try await httpClient.performRequest(
            method: .post,
            path: "/webhooks/paystack",
            headers: [
                "x-paystack-signature": paystackSignature
            ],
            requestOptions: requestOptions
        )
    }

    /// Receives payment event notifications from Flutterwave. Verified using the verif-hash header against the configured encryption key.
    ///
    /// - Parameter verifHash: Flutterwave verification hash
    /// - Parameter requestOptions: Additional options for configuring the request, such as custom headers or timeout settings.
    public func webhooksControllerFlutterwave(verifHash: String? = nil, requestOptions: RequestOptions? = nil) async throws -> Void {
        return try await httpClient.performRequest(
            method: .post,
            path: "/webhooks/flutterwave",
            headers: [
                "verif-hash": verifHash
            ],
            requestOptions: requestOptions
        )
    }

    /// Receives payment callback notifications from DPO Group (DirectPay Online). Verifies the transaction token status and updates payment accordingly.
    ///
    /// - Parameter requestOptions: Additional options for configuring the request, such as custom headers or timeout settings.
    public func webhooksControllerDpo(requestOptions: RequestOptions? = nil) async throws -> Void {
        return try await httpClient.performRequest(
            method: .post,
            path: "/webhooks/dpo",
            requestOptions: requestOptions
        )
    }

    /// Receives Instant Payment Notifications (IPN) from PayU South Africa. Updates payment status based on the transaction state.
    ///
    /// - Parameter requestOptions: Additional options for configuring the request, such as custom headers or timeout settings.
    public func webhooksControllerPayu(requestOptions: RequestOptions? = nil) async throws -> Void {
        return try await httpClient.performRequest(
            method: .post,
            path: "/webhooks/payu",
            requestOptions: requestOptions
        )
    }

    /// Receives IPN (Instant Payment Notification) callbacks from Pesapal. Fetches transaction status using the OrderTrackingId and updates payment.
    ///
    /// - Parameter requestOptions: Additional options for configuring the request, such as custom headers or timeout settings.
    public func webhooksControllerPesapal(requestOptions: RequestOptions? = nil) async throws -> Void {
        return try await httpClient.performRequest(
            method: .post,
            path: "/webhooks/pesapal",
            requestOptions: requestOptions
        )
    }

    /// Receives event notifications from Stripe (e.g. checkout.session.completed, payment_intent.succeeded). Verified using the stripe-signature header with the configured webhook secret.
    ///
    /// - Parameter stripeSignature: Stripe webhook signature
    /// - Parameter requestOptions: Additional options for configuring the request, such as custom headers or timeout settings.
    public func webhooksControllerStripe(stripeSignature: String, requestOptions: RequestOptions? = nil) async throws -> Void {
        return try await httpClient.performRequest(
            method: .post,
            path: "/webhooks/stripe",
            headers: [
                "stripe-signature": stripeSignature
            ],
            requestOptions: requestOptions
        )
    }
}